File: kotlin/lesson 6/PreOrder.kt
// https://leetcode.com/problems/binary-tree-preorder-traversal/

class Solution {
    fun preorderTraversal(root: TreeNode?): List<Int> {
        val result = mutableListOf<Int>()
        preOrder(root, result)
        return result
    }
    
    fun preOrder(node: TreeNode?, result: MutableList<Int>) {
        if (node == null) return

        result.add(node.`val`)
        preOrder(node.left, result)
        preOrder(node.right, result)
    }
}

File: ruby/lesson 6/pre_order.rb
# https://leetcode.com/problems/binary-tree-preorder-traversal/

def preorder_traversal(root)
    result = []
    preorder(root, result)
    result
end

def preorder(node, result)
    return if node.nil?

    result.push(node.val)
    preorder(node.left, result)
    preorder(node.right, result)
end